quickfixj-core/src/main/java/quickfix/mina/WatermarkTracker.java

```diff
@@ -0,0 +1,185 @@
+ * ALv2 (http://www.apache.org/licenses/LICENSE-2.0)
+ */
+
+package quickfix.mina;
+
+import java.util.AbstractCollection;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * A blocking queue wrapper implementing watermarks-based back pressure propagation
+ * from the queue sink to one or more logical sources.
+ *
+ * @param <E> payload type
+ * @param <S> logical source key type
+ *
+ * @author Vladimir Lysyy (mrbald@github)
+ */
+public class WatermarkTracker<E,S> implements QueueTracker<E> {
+    private final BlockingQueue<E> queue;
+    private final long lowerWatermark;
+    private final long upperWatermark;
+    private final Consumer<S> onLowerWatermarkCrossed;
+    private final Consumer<S> onUpperWatermarkCrossed;
+    private final Function<E, S> classifier;
+    private final Function<S, StreamTracker> trackerSupplier;
+
+    class StreamTracker {
+        private final S key;
+        long counter = 0;
+        private boolean suspended = false;
+
+        StreamTracker(S key) {
+            this.key = key;
+        }
+
+        synchronized void incoming(int n) {
+            if ((counter += n) >= upperWatermark && !suspended) {
+                suspended = true;
+                onUpperWatermarkCrossed.accept(key);
+            }
+        }
+
+        synchronized void outgoing(int n) {
+            if ((counter -= n) == lowerWatermark && suspended) {
+                suspended = false;
+                onLowerWatermarkCrossed.accept(key);
+            }
+        }
+
+        synchronized boolean isSuspended() {
+            return suspended;
+        }
+    }
+
+    public static <E, Void> WatermarkTracker<E, Void> newMono(
+            BlockingQueue<E> queue,
+            long lowerWatermark, long upperWatermark,
+            Runnable onLowerWatermarkCrossed, Runnable onUpperWatermarkCrossed) {
+        return new WatermarkTracker<E, Void>(queue, lowerWatermark, upperWatermark, onLowerWatermarkCrossed, onUpperWatermarkCrossed);
+    }
+
+    public static <E, S> WatermarkTracker<E, S> newMulti(
+            BlockingQueue<E> queue,
+            long lowerWatermark, long upperWatermark,
+            Function<E, S> classifier,
+            Consumer<S> onLowerWatermarkCrossed, Consumer<S> onUpperWatermarkCrossed) {
+        return new WatermarkTracker<>(queue, lowerWatermark, upperWatermark, classifier, onLowerWatermarkCrossed, onUpperWatermarkCrossed);
+    }
+
+    private WatermarkTracker(
+            BlockingQueue<E> queue,
+            long lowerWatermark, long upperWatermark,
+            Function<E, S> classifier,
+            Consumer<S> onLowerWatermarkCrossed, Consumer<S> onUpperWatermarkCrossed) {
+        assert lowerWatermark >= 0 && lowerWatermark < upperWatermark;
```

Comment:
I am not a fan of assertions in non-test code. Assertions are disabled by default anyway. Could you please remove them? Thanks

Reply:
It's more of a documenting assertion (as you said, they are disabled in prod). I'll remove in the next commit.